action-server/covidflow/actions/test_navigation_form.py

```diff
@@ -0,0 +1,245 @@
+        dispatcher.utter_message(
+            attachment=_locations_carousel(
+                tracker.get_slot(LANGUAGE_SLOT), domain, testing_locations
+            )
+        )
+        return {
+            POSTAL_CODE_SLOT: postal_code,
+            LOCATIONS_SLOT: [location.raw_data for location in testing_locations],
+        }
+
+    def validate_test_navigation__try_different_address(
+        self,
+        value: bool,
+        dispatcher: CollectingDispatcher,
+        tracker: Tracker,
+        domain: Dict[Text, Any],
+    ) -> Dict[Text, Any]:
+
+        if value:
+            return {POSTAL_CODE_SLOT: None, LOCATIONS_SLOT: None}
+
+        dispatcher.utter_message(template="utter_test_navigation__acknowledge")
+        return {}
+
+    def submit(
+        self,
+        dispatcher: CollectingDispatcher,
+        tracker: Tracker,
+        domain: Dict[Text, Any],
+    ) -> List[Dict]:
+        if tracker.get_slot(LOCATIONS_SLOT):
+            success = True
+        else:
+            success = False
+
+        return [SlotSet(TEST_NAVIGATION_SUCCESS_SLOT, success)]
+
+
+def _get_postal_code(text: str) -> Optional[str]:
+    matches = POSTAL_CODE_REGEX.search(text)
+    if matches:
+        return matches.group()
+    return None
+
+
+def _check_postal_code_error_counter(
+    tracker: Tracker, dispatcher: CollectingDispatcher
+) -> Dict[Text, Any]:
+    try_counter = tracker.get_slot(INVALID_POSTAL_CODE_COUNTER_SLOT)
+
+    if try_counter == MAX_POSTAL_CODE_TRIES:
+        dispatcher.utter_message(
+            template="utter_test_navigation__invalid_postal_code_max"
+        )
+        return {POSTAL_CODE_SLOT: None, END_FORM_SLOT: True}
+
+    dispatcher.utter_message(template="utter_test_navigation__invalid_postal_code")
+
+    return {
+        POSTAL_CODE_SLOT: None,
+        INVALID_POSTAL_CODE_COUNTER_SLOT: try_counter + 1,
+    }
+
+
+def _locations_carousel(
+    language: str, domain: Dict[Text, Any], testing_locations: List[TestingLocation]
+) -> Dict[Text, Any]:
+    titles_response = domain.get("responses", {}).get(
+        "utter_test_navigation__display_titles", []
+    )
+
+    titles = titles_response[0].get("custom", {}) if len(titles_response) >= 1 else {}
+    cards = [
+        _generate_location_card(location, titles, language)
+        for location in testing_locations
+    ]
+    return {
+        "type": "template",
+        "payload": {"template_type": "generic", "elements": cards},
+    }
+
+
+def _generate_location_card(
+    location: TestingLocation, titles: Dict[Text, Text], language: str
+) -> Dict[Text, Any]:
+    phone_number = (
+        _format_phone_number(location.phones[0]) if len(location.phones) >= 1 else None
+    )
+    buttons = (
+        [
+            {
+                "title": f"{titles.get('call_button')}{phone_number['full']}",
+                "type": "web_url",
+                "url": f"tel:{phone_number['link']}",
```

Comment:
there's a lot of strings in this that are repeated and could be extracted into constants